### PR TITLE
[torch.distributed][DDP] Disable DDP bucketing for the first iteration

### PR DESCRIPTION
**Pull request by xush6528**
Summary:
# [Debug Story] Training Hanging and DDP Bucketing

**What are the characteristics of the hanging training instance?**

The model uses TorchRec `PooledEmbeddingArch` and corresponding sharding solution.

The model config difference to trigger this hanging issue is turning on position weighted embedding tables.

A feature processor module, `GroupedPositionWeightedModule`, is constructed on all ranks, but `GroupedPositionWeightedModule.foward(...)` is only [called on subset ranks of the whole world](https://fburl.com/code/yqrmtvli).

**What was the initial manifested error?**

The training was stuck in the first iteration.

**What are useful debugging tools this time?**

After turning off [static_graph in DDP](https://fburl.com/code/4io81p5i), we saw there were sparse feature lengths becoming negative values after all-to-all collectives. Hanging becomes fatal failure.

After turning on [torch.distributed DETAIL debugging mode](https://fburl.com/code/cp8e28mm), we saw 2 trainers sent out mismatched collectives, one doing all-to-all, the other doing all-reduce. So we know the negative values comes from all-to-all being matched with all-reduce. the error had happened ahead, which is the wrong timing of either doing all-reduce or all-to-all.

With more added loggings inside of DDP, it turned out the DDP decided to do all-reduce at different timings across different ranks.

**What is DDP bucketing?**

Once a gradient is ready on a rank, DDP uses all-reduce to synchronize the average of this gradient across all ranks.

Say we have 4 tensor ops. A, B, C, D.

In the most naive version, we could do one synchronization when all gradients in the full backward graph are ready.

The time sequence would be,

* D.grad
* C.grad
* B.grad
* A.grad
* All reduce on [D.grad, C.grad, B.grad, A.grad].

But that would be a huge waste of communication channel bandwidth.

With DDP bucketing, we could put ahead some gradient synchronization batch by batch. The above time sequence now becomes,

* D.grad
* C.grad
* All reduce on [D.grad, C.grad].
* B.grad
* A.grad
* All reduce on [B.grad, A.grad].

With gradient computation overlaps with communication, bucketing technique brings better DDP execution performance.

**What exactly went wrong in this case?**

1. The bucketing doesn’t honor backward graph execution order.
2. There are other collectives comm ops in backward graph.
3. There are unused parameters (i.e unused sub-module) in subset ranks of the whole world.

Using the above example again, we have 4 tensor ops. A, B, C, D.

Say we have 2 trainers,

B is the feature processor module.

B only runs on trainer 0 (both forward and backward), but not on trainer1.

C is the All-to-all (Pooled embeddings distribution).

C sends out all-to-all collective in both its forward and backward pass.

Keep assuming all other ops run on both trainers.

trainer_0 op sequence is,

A, B (feature preproc), C (all-to-all), D | D.grad, C.grad (reverse all-to-all), B.grad (feature proc grads), A.grad

trainer_1 op sequence is,

A, C (all-to-all), D | D.grad, C.grad (reverse all-to-all), A.grad

Even though the correct bucketing should be (same bucketing for both ranks),

* bucket_0, [D.grad, C.grad]
* bucket_1, [B.grad, A.grad]

but because of 1), they end up like,

* bucket_0, [B.grad, D.grad]
* bucket_1, [C.grad, A.grad]

Plus 2) and 3), the time sequence could like,

(check mark represents the gradient is ready)

(bucket is ready to do synchronization if all its enclosing gradients are ready)

* trainer_0
   * t0,
      * D.grad
      * bucket_0, [B.grad, D.grad ✓]
   * t1,
      * **C.grad all-to-all**
      * C.grad ✓
      * bucket_1, [C.grad ✓, A.grad]
   * t2
      * B.grad
      * bucket_0, [B.grad ✓, D.grad ✓] ✓
   * t3
      * All-reduce for bucket_0
   * t4
      * A.grad
      * bucket_1, [C.grad ✓, A.grad ✓] ✓
* trainer_1
   * t0,
      * D.grad
      * bucket_0, [B.grad ✓, D.grad ✓] ✓. (Because B is not used on trainer_1, DDP marks its gradient as ready immediately.)
   * t1,
      * **All-reduce for bucket_0**
   * t2
      * C.grad all-to-all
      * bucket_1, [C.grad ✓, A.grad]
   * t3
      * A.grad
      * bucket_1, [C.grad ✓, A.grad ✓] ✓

This is why trainer_0 all-to-all is matched up with trainer_1 all-reduce.

**What is the solution for fixing DDP?**

Disable DDP bucketing for the first iteration. D34051938

This is because after the first iteration, buckets will be built again based on real backward graph execution order.

So the slow gradient synchronization only affects the first iteration.

Test Plan: tbd

Differential Revision: D34051938

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
Great debugging and root cause this issue!

regarding the fix, rebuilt bucket is only called when find_unused_parameters=False or when static graph is enabled. Because when find_unused_parameters=False, grad execution order could change every iteration.

there are two ways to further fix this:
1. enable rebuilt bucket even when find_unused_parameters=False
2. enable static graph for PyPer, make sure the second iteration use rebuilt bucket in static graph mode. right now, static graph uses rebuilt bucket at third iteration.

**Review by rohan-varma — COMMENTED** [review]
Great fix! Although I don't know that we should disable bucketing for the first iteration for all workflows. It seems like we can use DDP communication hook to get what we want here? 

```
tensors = []
def delay_allreduce_hook(bucket):
    tensors.extend(bucket.buffer())
    if bucket.is_last():
        dist.all_reduce(torch.cat(tensors))

def disable_bucketing_until_rebuilt(bucket) -> Callable:
    return delay_allreduce_hook if not reducer.rebuilt() else allreduce_hook()
```

**Comment by zhaojuanmao** [comment]
> Great fix! Although I don't know that we should disable bucketing for the first iteration for all workflows. It seems like we can use DDP communication hook to get what we want here?
> 
> ```
> tensors = []
> def delay_allreduce_hook(bucket):
>     tensors.extend(bucket.buffer())
>     if bucket.is_last():
>         dist.all_reduce(torch.cat(tensors))
> 
> def disable_bucketing_until_rebuilt(bucket) -> Callable:
>     return delay_allreduce_hook if not reducer.rebuilt() else allreduce_hook()
> ```

delay_all_reduce right now does not rebuilt_bucket, so the first iteration after delay_all_reduce will still use the static bucket.  To use delay_all_reduce, we need to make delay_all_reduce path rebuilt bucket first.

**Comment by rohan-varma** [comment]
> > Great fix! Although I don't know that we should disable bucketing for the first iteration for all workflows. It seems like we can use DDP communication hook to get what we want here?
> > ```
> > tensors = []
> > def delay_allreduce_hook(bucket):
> >     tensors.extend(bucket.buffer())
> >     if bucket.is_last():
> >         dist.all_reduce(torch.cat(tensors))
> > 
> > def disable_bucketing_until_rebuilt(bucket) -> Callable:
> >     return delay_allreduce_hook if not reducer.rebuilt() else allreduce_hook()
> > ```
> 
> delay_all_reduce right now does not rebuilt_bucket, so the first iteration after delay_all_reduce will still use the static bucket. To use delay_all_reduce, we need to make delay_all_reduce path rebuilt bucket first.

Oh I wasn't referring to the current implementation of delay allreduce in C++, but rather as a communication hook. If we do this hook based approach, the bucketing/rebuilding is completely separate and DDP will just call the hook after buckets are set up appropriately.

**Comment by zhaojuanmao** [comment]
> > > Great fix! Although I don't know that we should disable bucketing for the first iteration for all workflows. It seems like we can use DDP communication hook to get what we want here?
> > > ```
> > > tensors = []
> > > def delay_allreduce_hook(bucket):
> > >     tensors.extend(bucket.buffer())
> > >     if bucket.is_last():
> > >         dist.all_reduce(torch.cat(tensors))
> > > 
> > > def disable_bucketing_until_rebuilt(bucket) -> Callable:
> > >     return delay_allreduce_hook if not reducer.rebuilt() else allreduce_hook()
> > > model.register_comm_hook(disable_bucketing_until_rebuilt)
> > > ```
> > 
> > 
> > delay_all_reduce right now does not rebuilt_bucket, so the first iteration after delay_all_reduce will still use the static bucket. To use delay_all_reduce, we need to make delay_all_reduce path rebuilt bucket first.
> 
> Oh I wasn't referring to the current implementation of delay allreduce in C++, but rather as a communication hook. If we do this hook based approach, the bucketing/rebuilding is completely separate and DDP will just call the hook after buckets are set up appropriately.

I see, but the first iteration after delay_all_reduce() will still use the static bucket, right? grad ready order is not recorded in the delay_all_reduce(), it can not be rebuilt in the first iteration after delay_all_reduce()

**Comment by rohan-varma** [comment]
> I see, but the first iteration after delay_all_reduce() will still use the static bucket, right? grad ready order is not recorded in the delay_all_reduce(), it can not be rebuilt in the first iteration after delay_all_reduce()

Reducer should pass the grad buckets in the correct order to python: 

https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L841

https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L851

**Comment by zhaojuanmao** [comment]
> > I see, but the first iteration after delay_all_reduce() will still use the static bucket, right? grad ready order is not recorded in the delay_all_reduce(), it can not be rebuilt in the first iteration after delay_all_reduce()
> 
> Reducer should pass the grad buckets in the correct order to python, unless I'm missing something. i.e. communication hook should be independent of bucketing:
> 
> https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L841
> 
> https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L851
> 
> i.e. grad ready order is still recorded, we just call into the comm hook when needed.

You are right! I thought the comm hook delay_all_reduce is a separate path that is like the _delay_all_reduce in static graph.  In this case, comm hook approach will work as well!. 

But it will be a burden to let each use case add a comm hook, right? feel native support/fix in DDP for these use cases is better

**Comment by rohan-varma** [comment]
Can probably land this one after https://github.com/pytorch/pytorch/pull/73276#pullrequestreview-891310443 lands cc @zhaojuanmao

**Comment by zhaojuanmao** [comment]
@xush6528 #73276 was landed, would you please rebase to latest master and see whether tests will pass, thanks for fixing this!

**Comment by pritamdamania87** [comment]
@rohan-varma @zhaojuanmao Just came across this PR and the associated discussion regarding fixing this issue. My understanding is that with this PR we now never build buckets in the first iteration and always rebuild buckets after the first iteration (for all combinations of static_graph, find_unused_parameters etc.)? If so, can we ensure we add enough test coverage to ensure we always rebuild buckets after the first iteration?

**Comment by pritamdamania87** [comment]
Also, what is the configuration of DDP for the issue @xush6528 described in the PR summary? Is static_graph=False and find_unused_parameters=True?

**Comment by zhaojuanmao** [comment]
> @rohan-varma @zhaojuanmao Just came across this PR and the associated discussion regarding fixing this issue. My understanding is that with this PR we now never build buckets in the first iteration and always rebuild buckets after the first iteration (for all combinations of static_graph, find_unused_parameters etc.)? If so, can we ensure we add enough test coverage to ensure we always rebuild buckets after the first iteration?

@pritamdamania87 the purpose of #73276 is to make sure rebuild buckets after first iteration for all combinations of configures

File: torch/nn/parallel/distributed.py
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:710]: let's add a condition for this:  if static_graph=True or find_unused_parameter=False, use [sys.maxsize]; otherwise keep [dist._DEFAULT_FIRST_BUCKET_BYTES, self.bucket_bytes_cap]. 

once #73732 is landed, we can remove this condition

**Review by zhaojuanmao — COMMENTED** [review]
add one more comment to unblock this PR, also please address the test failures. Thanks for fixing this!

**Review by zhaojuanmao — APPROVED** [review]
looks good!

**Pull request closed without merging**
